sdk/core/azure-core/src/main/java/com/azure/core/util/ClientOptions.java

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.annotation.Fluent;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Client Options for setting common properties for example applicationId. Most of these properties are applied on
+ * request being send to Azure Service but some could be used for other purpose also for example applicationId could be
+ * used for telemetry.
+ */
+@Fluent
+public final class ClientOptions {
+    private final ClientLogger logger = new ClientLogger(ClientOptions.class);
+    private final Map<String, Header> headers = new ConcurrentHashMap<>();
+
+    private static final int MAX_APPLICATION_ID_LENGTH = 24;
+    private String applicationId;
+
+    /**
+     * Gets the applicationId.
+     *
+     * @return The applicationId.
+     */
+    public String getApplicationId() {
+        return applicationId;
+    }
+
+    /**
+     * Sets the applicationId provided.
+     *
+     * @param applicationId to be set.
+     * @return updated {@link ClientOptions}.
+     *
+     */
+    public ClientOptions setApplicationId(String applicationId) {
+        if (!CoreUtils.isNullOrEmpty(applicationId)) {
```

Comment:
nit: I'm not a fan of multiple nested if statements because it makes things harder to read if it's not necessary. Can we bounce early?

```java
if (CoreUtils.isNullorEmpty(applicationId)) { 
    return this;
}
... 
// other stuff.
```